SPIFlashLogger.device.lib.nut

```diff
@@ -0,0 +1,956 @@
+// Indicates that sector was closed on writing
+const SECTOR_WRITE_DONE = 0x04;
+// Indicates that sector has at least one start code
+const SECTOR_HAS_START_CODE = 0x08;
+// indicates that start code could be invalid
+// because object was erased
+const SECTOR_HAS_REMOVED_OBJECT = 0x08;
+
+class SPIFlashLogger {
+
+    static version = "3.0.0";
+
+    _flash = null; // hardware.spiflash or an object with an equivalent interface
+    _serializer = null; // github.com/electricimp/serializer (or an object with an equivalent interface)
+
+    _size = null; // The size of the spiflash
+    _start = null; // First block to use for logging
+    _end = null; // Last block to use for logging
+    _len = null; // The length of the flash available (end-start)
+    _sectors = 0; // The number of sectors in _len
+    _max_data = 0; // The maximum data we can push at once
+
+    _at_sec = 0; // Current sector we're writing to
+    _at_pos = 0; // Current position we're writing to in the sector
+
+    _enables = 0; // Counting semaphore for _enable/_disable
+    _sectorsMap = null; // list of sector objects
+
+    // This calss uses to read the serialized object from the flash
+    // and de-serialize it.
+    _loggerSerializedObject = class {
+        _flash = null;
+        _logger = null;
+        _addr = null;
+        _isValid = false;
+        _payload = null;
+        _len = 0;
+
+        // constructor
+        // Parameters:
+        //      addr - object address
+        //      logger - the SPIFlashLogger object
+        //      flash - the spiflash object
+        constructor(addr, logger, flash) {
+            _logger = logger;
+            _flash = flash;
+            _addr = addr;
+
+            _isValid = false;
+            _payload = null;
+            _len = 0;
+
+            init(addr);
+        }
+
+        // internal method to initialize object:
+        // check start code and read object size
+        function init(pos) {
+            local requested_pos = pos;
+            _logger._enable();
+            // Get the meta (for checking) and the object length (to know how much to read)
+            local marker = _flash.read(pos, SPIFLASHLOGGER_OBJECT_MARKER_SIZE).tostring();
+            local len = _flash.read(pos + SPIFLASHLOGGER_OBJECT_MARKER_SIZE, 2).readn('w');
+            _logger._disable();
+            _len = len;
+            _isValid = (marker == SPIFLASHLOGGER_OBJECT_MARKER && len > 0 && len < _logger._size);
+        }
+
+        //
+        // Indicates if  object has valid length and startcode
+        function isValid() {
+            return _isValid;
+        }
+
+        //
+        // Return object address
+        //
+        function getAddress() {
+            return _addr;
+        }
+
+        //
+        // De-serialize object data and return payload
+        //
+        function getPayload() {
+            if (!_isValid)
+                return null;
+
+            if (_payload != null)
```

Comment:
What happens if the content on the SPI flash changed after it was cached here?

Reply:
if _payload was cached here then the same object will be returned;

if object is not valid anymore, but _isValid is true, then we should fail on deserialization and return null as payload.  For the read() method it is lead to the onFinish() call.  For the readSync() we just return defaultValue according to specification